core/tests/test-dependencies/typescript-ast-utilities.ts

```diff
@@ -0,0 +1,212 @@
+ */
+
+import path from 'path';
+import fs from 'fs';
+import {Decorator, Node, Project, SourceFile, ts} from 'ts-morph';
+
+/**
+ * Gets the root directory of the project by traversing up the directory tree
+ * until a package.json file is found.
+ */
+export const ROOT_DIRECTORY = (() => {
+  let current = __dirname;
+  while (!fs.existsSync(path.join(current, 'package.json'))) {
+    current = path.dirname(current);
+  }
+  return current;
+})();
+
+export const project = new Project({
+  tsConfigFilePath: path.join(ROOT_DIRECTORY, 'tsconfig.json'),
+  skipFileDependencyResolution: true,
+});
+
+export enum AngularDecorators {
+  Component = 'Component',
+  Directive = 'Directive',
+  Pipe = 'Pipe',
+  Module = 'NgModule',
+}
+
+type ResolvedModuleWithFailedLookupLocations =
+  ts.ResolvedModuleWithFailedLookupLocations & {
+    failedLookupLocations: string[];
+  };
+
+/**
+ * List of types that are defined in node.
+ */
+const NODE_TYPE_ROOTS = ['fs', 'console', 'path', 'child_process'];
+
+/**
+ * Checks if a module is a node module.
+ */
+export const isNodeModule = (modulePath: string): boolean => {
+  return modulePath.includes('node_modules');
+};
+
+/**
+ * Gets the potential module lookup locations using the TypeScript compiler.
+ */
+const getPotentialModuleLookupLocations = (
+  modulePath: string,
+  containingFile: string
+): string[] => {
+  const failedLookup = ts.resolveModuleName(
+    modulePath,
+    containingFile,
+    project.getCompilerOptions(),
+    project.getModuleResolutionHost()
+  ) as ResolvedModuleWithFailedLookupLocations;
+
+  return failedLookup.failedLookupLocations
+    .filter(location => location.endsWith('.ts'))
+    .map(location => location.replace('.ts', ''));
+};
+
+/**
+ * Fallback module resolution when TypeScript compiler fails to resolve a module.
+ * Looks through potential module lookup locations and returns the location where
+ * the module is found.
+ */
+const fallbackResolveModule = (
+  modulePath: string,
+  containingFile: string
+): ts.ResolvedModuleFull => {
+  const potentialModuleLookupLocations = getPotentialModuleLookupLocations(
+    modulePath,
+    containingFile
+  );
+
+  for (const location of potentialModuleLookupLocations) {
+    if (fs.existsSync(location)) {
+      return {
+        resolvedFileName: location,
+        extension: path.extname(location),
+      };
+    }
+  }
+  throw new Error(
+    `Failed to resolve module ${modulePath} at ${containingFile}.`
+  );
+};
+
+/**
+ * Resolves a module path with Webpack resolution.
+ */
+const resolveModuleUsingWebpack = (
+  modulePath: string,
+  containingFile: string
+): ts.ResolvedModuleFull => {
+  const resolved = ts.resolveModuleName(
+    'core/templates/' + modulePath,
+    containingFile,
+    project.getCompilerOptions(),
+    project.getModuleResolutionHost()
+  );
+  if (resolved.resolvedModule === undefined) {
+    return fallbackResolveModule(modulePath, containingFile);
+  }
```

Comment:
When will this happen? Should we log this?

Reply:
This will happen usually when the TypeScript compiler isn't able to resolve it, which occurs most times with any extension that is not .ts. So I don't think we need to log here, but the fallbackResolveModule will error when the module cannot be completely resolved.